
This patch redirects flutter_tools' Gradle project to build in a writable
directory at `$HOME/.cache/flutter/nix-flutter-tools-gradle/$engineShortRev`,
instead of within the read-only Nix Store.

Without this patch, flutter_tools' Gradle project tries to generate `build` and `.gradle`
directories within the Nix Store. Resulting in read-only errors when trying to build a
Flutter Android app at runtime. In Gradle 9+, even the project directory itself must be
writable.

This patch takes advantage of the fact settings.gradle takes priority over settings.gradle.kts.
When a Flutter app runs `includeBuild("${settings.ext.flutterSdkPath}/packages/flutter_tools/gradle")`,
Gradle evaluates the `settings.gradle` we add, which:
1. Creates symlinks to `build.gradle.kts` and `src/` in the writable cache directory
2. Sets `rootProject.projectDir` to the writable cache directory
3. Applies the original `settings.gradle.kts` for dependency resolution
4. Redirects `buildDirectory` to the cache

To move `.gradle` as well, the `--project-cache-dir` argument must be passed to the Gradle wrapper.
Changing the `GradleUtils.getExecutable` function signature is a deliberate choice, to ensure that no new unpatched usages slip in.
diff --git a/packages/flutter_tools/gradle/settings.gradle b/packages/flutter_tools/gradle/settings.gradle
new file mode 100644
index 0000000..4bb7e8e
--- /dev/null
+++ b/packages/flutter_tools/gradle/settings.gradle
@@ -0,0 +1,22 @@
+def engineShortRev = (new File("$settingsDir/../../../bin/internal/engine.version")).text.take(10)
+def dir = new File("$System.env.HOME/.cache/flutter/nix-flutter-tools-gradle/$engineShortRev")
+dir.mkdirs()
+
+def buildScript = new File(dir, "build.gradle.kts")
+if (!buildScript.exists()) {
+  java.nio.file.Files.createSymbolicLink(buildScript.toPath(), new File("$settingsDir/build.gradle.kts").toPath())
+}
+def srcLink = new File(dir, "src")
+if (!srcLink.exists()) {
+  java.nio.file.Files.createSymbolicLink(srcLink.toPath(), new File("$settingsDir/src").toPath())
+}
+
+rootProject.projectDir = dir
+
+apply from: new File("$settingsDir/settings.gradle.kts")
+
+gradle.allprojects { project ->
+  project.beforeEvaluate {
+    project.layout.buildDirectory = new File("$dir/build")
+  }
+}
diff --git a/packages/flutter_tools/gradle/build.gradle.kts b/packages/flutter_tools/gradle/build.gradle.kts
index 45065e0..6dd177f 100644
--- a/packages/flutter_tools/gradle/build.gradle.kts
+++ b/packages/flutter_tools/gradle/build.gradle.kts
@@ -4,6 +4,11 @@
 
 import org.jetbrains.kotlin.gradle.dsl.JvmTarget
 
+// While flutter_tools runs Gradle with a --project-cache-dir, this startParameter
+// is not passed correctly to the Kotlin Gradle plugin for some reason, and so
+// must be set here as well.
+gradle.startParameter.projectCacheDir = layout.buildDirectory.dir("cache").get().asFile
+
 plugins {
     `java-gradle-plugin`
     groovy
diff --git a/packages/flutter_tools/lib/src/android/gradle.dart b/packages/flutter_tools/lib/src/android/gradle.dart
index b083bca..cb6fb8d 100644
--- a/packages/flutter_tools/lib/src/android/gradle.dart
+++ b/packages/flutter_tools/lib/src/android/gradle.dart
@@ -545,9 +545,9 @@ To fix this, you can either:
       logger: _logger,
     );
 
-    final options = <String>[];
-
-    final String gradleExecutablePath = _gradleUtils.getExecutable(project);
+    final [String gradleExecutablePath, ...List<String> options] = _gradleUtils.getExecutable(
+      project,
+    );
 
     // All automatically created files should exist.
     if (configOnly) {
@@ -883,7 +883,7 @@ To fix this, you can either:
       'aar_init_script.gradle',
     );
     final command = <String>[
-      _gradleUtils.getExecutable(project),
+      ..._gradleUtils.getExecutable(project),
       '-I=$initScript',
       '-Pflutter-root=$flutterRoot',
       '-Poutput-dir=${outputDirectory.path}',
@@ -1000,6 +1000,10 @@ To fix this, you can either:
     final results = <String>[];
 
     try {
+      final [String gradleExecutablePath, ...List<String> options] = _gradleUtils.getExecutable(
+        project,
+      );
+
       exitCode = await _runGradleTask(
         _kBuildVariantTaskName,
         preRunTask: () {
@@ -1015,10 +1019,10 @@ To fix this, you can either:
             ),
           );
         },
-        options: const <String>['-q'],
+        options: <String>[...options, '-q'],
         project: project,
         localGradleErrors: gradleErrors,
-        gradleExecutablePath: _gradleUtils.getExecutable(project),
+        gradleExecutablePath: gradleExecutablePath,
         outputParser: (String line) {
           if (_kBuildVariantRegex.firstMatch(line) case final RegExpMatch match) {
             results.add(match.namedGroup(_kBuildVariantRegexGroupName)!);
@@ -1052,6 +1056,10 @@ To fix this, you can either:
     late Stopwatch sw;
     var exitCode = 1;
     try {
+      final [String gradleExecutablePath, ...List<String> options] = _gradleUtils.getExecutable(
+        project,
+      );
+
       exitCode = await _runGradleTask(
         taskName,
         preRunTask: () {
@@ -1067,10 +1075,10 @@ To fix this, you can either:
             ),
           );
         },
-        options: <String>['-q', '-PoutputPath=$outputPath'],
+        options: <String>[...options, '-q', '-PoutputPath=$outputPath'],
         project: project,
         localGradleErrors: gradleErrors,
-        gradleExecutablePath: _gradleUtils.getExecutable(project),
+        gradleExecutablePath: gradleExecutablePath,
       );
     } on Error catch (error) {
       _logger.printError(error.toString());
diff --git a/packages/flutter_tools/lib/src/android/gradle_errors.dart b/packages/flutter_tools/lib/src/android/gradle_errors.dart
index eaaff3f..1c382b6 100644
--- a/packages/flutter_tools/lib/src/android/gradle_errors.dart
+++ b/packages/flutter_tools/lib/src/android/gradle_errors.dart
@@ -231,7 +231,12 @@ final flavorUndefinedHandler = GradleHandledError(
   },
   handler: ({required String line, required FlutterProject project, required bool usesAndroidX}) async {
     final RunResult tasksRunResult = await globals.processUtils.run(
-      <String>[globals.gradleUtils!.getExecutable(project), 'app:tasks', '--all', '--console=auto'],
+      <String>[
+        ...globals.gradleUtils!.getExecutable(project),
+        'app:tasks',
+        '--all',
+        '--console=auto',
+      ],
       throwOnError: true,
       workingDirectory: project.android.hostAppGradleRoot.path,
       environment: globals.java?.gradleEnvironment,
diff --git a/packages/flutter_tools/lib/src/android/gradle_utils.dart b/packages/flutter_tools/lib/src/android/gradle_utils.dart
index b00ce8b..c4db4e8 100644
--- a/packages/flutter_tools/lib/src/android/gradle_utils.dart
+++ b/packages/flutter_tools/lib/src/android/gradle_utils.dart
@@ -3,6 +3,7 @@
 // found in the LICENSE file.
 
 import 'package:meta/meta.dart';
+import 'package:path/path.dart';
 import 'package:process/process.dart';
 import 'package:unified_analytics/unified_analytics.dart';
 
@@ -224,9 +225,30 @@ class GradleUtils {
   final Logger _logger;
   final OperatingSystemUtils _operatingSystemUtils;
 
+  List<String> get _requiredArguments {
+    final String baseDir = switch (_platform.environment['XDG_CACHE_HOME']) {
+      final String cacheHome => cacheHome,
+      _ => join(
+        _platform.environment['HOME'] ?? throwToolExit('No cache directory has been specified.'),
+        '.cache',
+      ),
+    };
+    final String cacheDir = join(
+      baseDir,
+      'flutter',
+      'nix-flutter-tools-gradle',
+      _cache.engineRevision.substring(0, 10),
+    );
+
+    return <String>[
+      '--project-cache-dir=${join(cacheDir, 'cache')}',
+      '-Pkotlin.project.persistent.dir=${join(cacheDir, 'kotlin')}',
+    ];
+  }
+
   /// Gets the Gradle executable path and prepares the Gradle project.
   /// This is the `gradlew` or `gradlew.bat` script in the `android/` directory.
-  String getExecutable(FlutterProject project) {
+  List<String> getExecutable(FlutterProject project) {
     final Directory androidDir = project.android.hostAppGradleRoot;
     injectGradleWrapperIfNeeded(androidDir);
 
@@ -237,7 +259,7 @@ class GradleUtils {
       // If the Gradle executable doesn't have execute permission,
       // then attempt to set it.
       _operatingSystemUtils.makeExecutable(gradle);
-      return gradle.absolute.path;
+      return <String>[gradle.absolute.path, ..._requiredArguments];
     }
     throwToolExit(
       'Unable to locate gradlew script. Please check that ${gradle.path} '
diff --git a/packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart b/packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart
index f3dc2a9..9e04d15 100644
--- a/packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart
+++ b/packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart
@@ -3463,8 +3463,8 @@ Gradle Crashed
 
 class FakeGradleUtils extends Fake implements GradleUtils {
   @override
-  String getExecutable(FlutterProject project) {
-    return 'gradlew';
+  List<String> getExecutable(FlutterProject project) {
+    return const <String>['gradlew'];
   }
 }
 
diff --git a/packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart b/packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart
index 30fa79f..ab922e6 100644
--- a/packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart
+++ b/packages/flutter_tools/test/general.shard/android/gradle_errors_test.dart
@@ -1811,8 +1811,8 @@ Platform fakePlatform(String name) {
 
 class FakeGradleUtils extends Fake implements GradleUtils {
   @override
-  String getExecutable(FlutterProject project) {
-    return 'gradlew';
+  List<String> getExecutable(FlutterProject project) {
+    return const <String>['gradlew'];
   }
 }
 
